src/server/req_preemptjob.c

```diff
@@ -85,34 +285,73 @@ extern time_t time_now;
 void
 req_preemptjobs(struct batch_request *preq)
 {
-	int			i = 0;
-	int			count = 0;
-	job			*pjob = NULL;
-	preempt_job_info	*ppj = NULL;
-	pbs_sched		*psched;
+	int i = 0;
+	int count = 0;
+	job *pjob = NULL;
+	preempt_job_info *ppj = NULL;
+	pbs_sched *psched;
 
 	preq->rq_reply.brp_code = 0;
 	count = preq->rq_ind.rq_preempt.count;
 	psched = find_sched_from_sock(preq->rq_conn);
 
+	preempt_total = preq->rq_ind.rq_preempt.count;
+
+	if (preempt_jobs_list == NULL) {
+		if ((preempt_jobs_list = calloc(sizeof(preempt_job_info), preempt_total)) == NULL) {
+			req_reject(PBSE_SYSTEM, 0, preq);
+			log_err(errno, __func__, "Unable to allocate memory");
+			return;
+		}
+	}
+
+	preq->rq_reply.brp_choice = BATCH_REPLY_CHOICE_PreemptJobs;
+	preq->rq_reply.brp_un.brp_preempt_jobs.count = preq->rq_ind.rq_preempt.count;
+
 	for (i = 0; i < count; i++) {
 		ppj = &(preq->rq_ind.rq_preempt.ppj_list[i]);
 		pjob = find_job(ppj->job_id);
+		/* The job is already out of the way. This must have happened after the scheduler
+		 * queried the universe and before it tried to preempt the jobs.
+		 * Regardless of the preempt_order, use the correct reply code to what
+		 * actually happened so the scheduler correctly handles the job.
+		 */
+		if (pjob == NULL) {
+			sprintf(preempt_jobs_list[preempt_index].job_id, "%s", ppj->job_id);
+			strcpy(preempt_jobs_list[preempt_index].order, "D");
+			preempt_index++;
+			job_preempt_fail(preq, ppj->job_id);
```

Comment:
why do we need to fail it? You are already setting the job_id and order in the lines above

Reply:
@bhroam how about this comment?